contracts/test/AllowanceTarget.t.sol

```diff
@@ -0,0 +1,134 @@
+
+    // effectively a "beforeEach" block
+    function setUp() public {
+        // Setup
+        allowanceTarget = new AllowanceTarget(address(this));
+
+        // Label addresses for easier debugging
+        vm.label(address(this), "TestingContract");
+        vm.label(address(allowanceTarget), "AllowanceTarget");
+    }
+
+    /*********************************
+     *          test: setup          *
+     *********************************/
+
+    function testSetupAllowance() public {
+        assertEq(allowanceTarget.spender(), address(this));
+    }
+
+    /*********************************
+     *  test: setSpenderWithTimelock *
+     *********************************/
+
+    function testCantSetSpenderWithTimelockByRandomEOA() public {
+        vm.prank(bob);
+        vm.expectRevert("AllowanceTarget: not the spender");
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+    }
+
+    function testCantSetSpenderWithTimelockWithInvalidAddress() public {
+        vm.expectRevert("AllowanceTarget: new spender not a contract");
+        allowanceTarget.setSpenderWithTimelock(bob);
+        //vm.expectRevert("");
+    }
+
+    function testCantSetSpenderWithTimelockIfInProgress() public {
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+        address mySpender = address(new MockStrategy());
+        vm.expectRevert("AllowanceTarget: SetSpender in progress");
+        allowanceTarget.setSpenderWithTimelock(mySpender);
+    }
+
+    function testSetSpenderWithTimelock() public {
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+        assertEq(allowanceTarget.newSpender(), newSpender);
+        assertEq(allowanceTarget.timelockExpirationTime(), block.timestamp + 1 days);
+    }
+
+    /*********************************
+     *  test: setSpenderWithTimelock *
+     *********************************/
+
+    function testCantCompleteSetSpenderBeforeSet() public {
+        vm.expectRevert("AllowanceTarget: no pending SetSpender");
+        allowanceTarget.completeSetSpender();
+    }
+
+    function testCantCompleteSetSpenderTooEarly() public {
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+        vm.expectRevert("AllowanceTarget: time lock not expired yet");
+        allowanceTarget.completeSetSpender();
+    }
+
+    function testCompleteSetSpender() public {
+        allowanceTarget.setSpenderWithTimelock(newSpender);
+        // fast forward
+        vm.warp(allowanceTarget.timelockExpirationTime());
+        allowanceTarget.completeSetSpender();
+    }
+
+    /*********************************
+     *         test: teardown        *
+     *********************************/
+
+    function testCantTeardownIfNotSpender() public {
+        vm.prank(bob);
+        vm.expectRevert("AllowanceTarget: not the spender");
+        allowanceTarget.teardown();
+    }
+
+    function testTeardown() public {
+        // Seems like the test contract itself have issue with receving ether.
```

Comment:
I think we can add a `receive()` function to test contract

Reply:
`selfdestruct` doesn't require `receive` or `fallback` function in fact. I tried adding `receive()` and still not working as expected. Not sure why but mock contract works fine.

Reply:
Got it.